packages/compass-import-export/src/modules/import.ts

```diff
@@ -353,63 +392,98 @@ export const startImport = () => {
+      importDone = true;
+      errorLogWriteStream?.close();
+      progressCallback.flush();
+    }
+
+    track('Import Completed', {
+      duration: Date.now() - startTime,
+      file_type: fileType,
+      all_fields: exclude.length === 0,
+      stop_on_error_selected: stopOnErrors,
+      number_of_docs: result.docsWritten,
+      success: true,
+      aborted: result.aborted,
+    });
 
-        log.info(mongoLogId(1001000082), 'Import', 'Import completed', {
-          ns,
+    log.info(mongoLogId(1001000082), 'Import', 'Import completed', {
+      ns,
+      docsWritten: result.docsWritten,
+      docsProcessed: result.docsProcessed,
+    });
+
+    if (result.aborted) {
+      showCancelledToast({
+        errors,
+        errorLogFilePath: errorLogFilePath,
+      });
+    } else {
+      if (errors.length > 0) {
+        showCompletedWithErrorsToast({
           docsWritten: result.docsWritten,
+          errors,
           docsProcessed: result.docsProcessed,
+          errorLogFilePath: errorLogFilePath,
         });
-
-        dispatch(onFinished(result.docsWritten, result.docsProcessed));
-
-        const payload = {
-          ns,
-          size: fileSize,
-          fileType,
+      } else {
+        showCompletedToast({
           docsWritten: result.docsWritten,
-          fileIsMultilineJSON,
-          delimiter,
-          ignoreBlanks,
-          stopOnErrors,
-          hasExcluded: exclude.length > 0,
-          hasTransformed: transform.length > 0,
-        };
-        dispatch(globalAppRegistryEmit('import-finished', payload));
-      })
-      .catch((err) => {
-        dispatch(onFinished(err.result.docsWritten, err.result.docsProcessed));
-
-        track('Import Completed', {
-          duration: Date.now() - startTime,
-          file_type: fileType,
-          all_fields: exclude.length === 0,
-          stop_on_error_selected: stopOnErrors,
-          number_of_docs: err.result.docsWritten,
-          success: !err,
         });
+      }
+    }
 
-        log.error(mongoLogId(1001000081), 'Import', 'Import failed', {
-          ns,
-          docsWritten: err.result.docsWritten,
-          error: err.message,
-        });
-        debug('Error while importing:', err.stack);
+    dispatch(
+      onFinished({
+        aborted: !!result.aborted,
+        docsWritten: result.docsWritten,
+        docsProcessed: result.docsProcessed,
+        errors,
+      })
+    );
 
-        return dispatch(onFailed(err));
-      });
+    const payload = {
+      ns,
+      size: fileSize,
+      fileType,
+      docsWritten: result.docsWritten,
+      fileIsMultilineJSON,
+      delimiter,
+      ignoreBlanks,
+      stopOnErrors,
+      hasExcluded: exclude.length > 0,
+      hasTransformed: transform.length > 0,
+    };
+
+    // Don't emit when the data service is disconnected.
+    if (dataService === state.dataService.dataService) {
+      dispatch(globalAppRegistryEmit('import-finished', payload));
+    }
```

Comment:
I'm not sure this will have an effect that expect it to have (there is still an error thrown when you disconnect during import and tooltip stays on the screen) for two reasons:

1. `state` is a reference to the state created at the very beginning of the import as is this `dataService` variable (at the very top of the action you have `dataService = state.dataService.dataService`). So even when if your state changes and new `dataService` is set in state, you are referencing stale state, this condition will always be true inside this action. This is why `redux-thunk` docs even have a section on [how to correctly access state in actions](https://redux.js.org/usage/writing-logic-thunks#accessing-state):

> Since the state is synchronously updated as soon as the reducers process an action, you can call getState after a dispatch to get the updated state.

2. even if the state was correctly accessed here, there is no code in this store that would actually clean up data service from the store when data service is disconnected. I see that you are emitting a `cancelImport` action when data service disconnects, but this event will not do anything about dataService being in the store. This is one of the reasons redux documentation suggests actions to be modelled around [application events that occur](https://redux.js.org/style-guide/#model-actions-as-events-not-setters) and not specific "multi-purpose" setters / getters. If you do it like that, it's much easier to reason about what is the application flow and how different parts of the store should react to this. In this case `dataServiceDisconnected` can handle cancelling any ongoing operations and reducer can react to this action and clean up the state, which would ensure that you don't have disconnected dataService there

Reply:
Yes - thanks!
Updated the `data-service-disconnected` event handler to emit a redux event that is handled by two places:
- The data service reducer is now cleaned up/reset when the disconnected event happens
- The import reducer now aborts and resets the abortControllers on the disconnected event.

Updated this check to check for the the active `dataService` from `getState` matching the dataService that was initially used. No more error on disconnect while an import is in progress.